# Actividad 4

## Practica

In [19]:
cadena = input('[+] Introduce loa números a sumar separados por un espacio:\n>>> ')
lista, suma = [float(i)  if not i.upper() == 'STOP' else 'STOP' for i in cadena.split()], float()
# Quitamos lo que sobra
if 'STOP' in lista:
    lista = lista[:lista.index('STOP')]
lista.sort()  # Ordenamos la lista
for i in lista:
    suma += i
    
print(f'[-] La suma de los valores introducidos es la siguiente: {suma}')
print(f'[-] La lista ordenada es: {lista}')

[-] La suma de los valores introducidos es la siguiente: 49.9
[-] La lista ordenada es: [4.0, 5.4, 8.5, 9.0, 23.0]


**Run 1**<br/>
[+] Introduce loa números a sumar separados por un espacio: 23 4 5.4 8.5 9.0<br/>
[-] La suma de los valores introducidos es la siguiente: 49.9 <br/>
[-] La lista ordenada es: [4.0, 5.4, 8.5, 9.0, 23.0]<br/>
**Run 2**<br/>
[+] Introduce loa números a sumar separados por un espacio: 23 4 5.4 8.5 stop 9.0<br/>
[-] La suma de los valores introducidos es la siguiente: 40.9<br/>
[-] La lista ordenada es: [4.0, 5.4, 8.5, 23.0]

************************************

## Pregunta 1

Analicemos el contenido por partes:<br/>
   * 'abc' quedará tal cual se muestra ya que no tiene operadores adicionales.
   * 'xy' * 2 = 'xyxy'
   * 'a' * 3 = 'aaa'
   * Además ya que python respeta el orden de las operaciones primero se realizarán las multiplicaciones y tras ello realizará la suma de todo, por lo que el resultado será **'abcxyxyaaa'**<br/>
   
Teniendo lo anterior en cuenta, la respuesta es **C)**

## Pregunta 2
Nuevamente, analizaremos el contenido por pasos tomando cada uno de los caracteres:
   * ch = 'Z':<br/>
   '-' + chr(ord('Z') - 1) * 2 --> chr(ord('Z')-1) = Y --> salida = '-' + 'Y' * 2 = '-YY'
   * ch = ch = 'Y':<br/>
   '-YY' + chr(ord('Y') - 1) * 2 --> chr(ord('Y')-1) = X --> salida = '-YY' + 'X' * 2 = '-YYXX'
   * ch = 'X':<br/>
   '-YYXX' + chr(ord('X') - 1) * 2 --> chr(ord('X')-1) = W --> salida = '-YYXX' + 'W' * 2 = '-YYXXWW'
   * Al igual que en la pregunta anterior, hay que tener en cuenta que python respeta el orden de los operandos por lo que primero resolverá los paréntesis, después las multiplicaciones y por último las sumas.

   Por tanto, la respuesta será **B)**


## Pregunta 3
Analicemos cada una de las líneas:
   * float('2.1'): Esto es una conversion de string a float y dado que el formato de a ('2.1') es adecuado para la conversión se realizará sin problemas por lo que b = 2.1
   * int(2.1): En este caso estamos realizando la conversión de float a int, operación perfectamente valida por lo que c = 2 ya que al hacer esta conversión python se queda con la parte entera.
   * int('2.4'): En esta ocasión se está tratando de hacer una conversión de string a int pero el string no tiene un formato adecuado por lo que el programa devolverá un error.
   * Las otras dos opciones no tiene sentido analizarlas ya que d no llegara a tener ningún valor pues el programa ha dado un error.

Por lo tanto, la respuesta es **D)**

## Pregunta 4
En esta ocasión hay varias cosas a tener en cuenta:
   * Al hacer mi_lista1 = mi_lista[:] estamos realizando una copia de la lista por lo que id(mi_lista) e id(mi_lista1) no son el mismo y los cambios realizados en una de las listas no afectan a la otra.
   * Al ordenar mi_lista obtendremos lo siguiente: ['AA', 'a', 'aa']. Esto es debido a que los valores en la tabla ASCII de las letras mayúsculas son menores por lo que 'AA' es menor que 'a'.<br/>

Teniendo todo los expuesto en cuenta, hay dor respuestas correctas, **B)** y **D)**.

## Pregunta 5
Analicemos paso por paso y siguiendo el orden de ejecución:
  * Primero, for j in range(0, 4) if j % 2 == 0, esta linea hará que se ejecute dos veces, es decir, generará dos listas.
  * Segundo, for i in range(4), esta linea generará una lista con cuatro elementos, en total tendremos una lista de dos listas con cuatro elementos cada una.
  * Por último, i if j > 1 else -1, con esta linea se establece el valor de i, concretamente, se establece que para todo j (derivado del primer for) menor o igual a 1 i == -1 y para todo j > 1 i == 0, 1, 2, 3, en función del for.

Teniendo todo ello en cuenta tenemos que mi_lista tendrá la siguiente pinta: [[-1, -1, -1, -1], [0, 1, 2, 3]] por lo que la respuesta será **B)**
